src/validations/passport-gated/index.ts

```diff
@@ -27,46 +49,145 @@ function hasValidIssuanceAndExpiration(credential, proposalTs) {
+  if (!stamp) {
+    console.log('[passport] Stamp not supported', stampId);
+    throw new Error('Stamp not supported');
+  }
+  return credentials.some((credential) =>
+    stamp.credentials.includes(credential)
+  );
+}
+
+async function validateStamps(
+  currentAddress: string,
+  operator: string,
+  proposalTs: string,
+  requiredStamps: Array<string> = []
+): Promise<boolean> {
+  if (requiredStamps.length === 0) return true;
+
+  const stampsResponse = await fetch(GET_PASSPORT_STAMPS_URI + currentAddress, {
+    headers
+  });
+  const stampsData = await stampsResponse.json();
+
+  if (!stampsData?.items) {
+    console.log('[passport] Stamps Unknown error', stampsData);
+    throw new Error('Unknown error');
+  }
+  if (stampsData.items.length === 0) return false;
+
+  // check expiration for all stamps
+  const validStamps = stampsData.items
+    .filter((stamp: any) =>
+      hasValidIssuanceAndExpiration(stamp.credential, proposalTs)
+    )
+    .map((stamp: any) => stamp.credential.credentialSubject.provider);
+
+  if (operator === 'AND') {
+    return requiredStamps.every((stampId) =>
+      hasStampCredential(stampId, validStamps)
+    );
+  } else if (operator === 'OR') {
+    return requiredStamps.some((stampId) =>
+      hasStampCredential(stampId, validStamps)
+    );
+  }
+  return false;
+}
+
+function evalPassportScore(scoreData: any, minimumThreshold = 0): boolean {
+  // scoreData.evidence?.type === 'ThresholdScoreCheck' -> Returned if using Boolean Unique Humanity Scorer (should not be used)
+  if (scoreData.evidence?.type === 'ThresholdScoreCheck') {
+    return (
+      Number(scoreData.evidence.rawScore) > Number(scoreData.evidence.threshold)
+    );
+  }
+  // scoreData.score -> Returned if using Unique Humanity Score
+  return Number(scoreData.score) > minimumThreshold;
+}
+
+async function validatePassportScore(
+  currentAddress: string,
+  scoreThreshold: number
+): Promise<boolean> {
+  // always hit the /submit-passport endpoint to get the latest passport score
+  const submittedPassport = await fetch(POST_SUBMIT_PASSPORT_URI, {
+    headers,
+    method: 'POST',
+    body: JSON.stringify({ address: currentAddress, scorer_id: SCORER_ID })
+  });
+  const submissionData =
+    submittedPassport.ok && (await submittedPassport.json());
+
+  if (!submittedPassport.ok) {
+    const reason = !SCORER_ID
+      ? 'SCORER_ID missing'
+      : submittedPassport.statusText;
+    console.log('[passport] Scorer error', reason);
+    throw new Error(`Scorer error: ${reason}`);
+  }
+
+  // Scorer done calculating passport score during submission
+  if (submittedPassport.ok && submissionData.status === 'DONE') {
+    return evalPassportScore(submissionData, scoreThreshold);
+  }
+
+  // Try to fetch Passport Score if still processing (submittedPassport.status === 'PROCESSING')
+  for (let i = 0; i < PASSPORT_SCORER_MAX_ATTEMPTS; i++) {
+    const scoreResponse = await fetch(GET_PASSPORT_SCORE_URI + currentAddress, {
```

Comment:
This API will return only DONE and PROCESSING? 🤔

Reply:
right. As it always hits the `/submit-passport` endpoint first, the `GET_PASSPORT_SCORE_URI` will be called only if the former returns `PROCESSING` status